File: backend/src/main/java/com/example/skillsim/config/ScoreDataLoader.java
package com.example.skillsim.config;

import com.example.skillsim.model.ScoreEffect;
import com.example.skillsim.model.ScoreSkill;
import com.example.skillsim.repository.ScoreSkillRepository;
import com.opencsv.CSVReader;
import com.opencsv.CSVReaderBuilder;
import com.opencsv.exceptions.CsvValidationException;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.Reader;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.boot.CommandLineRunner;
import org.springframework.core.io.ClassPathResource;
import org.springframework.stereotype.Component;

@Component
public class ScoreDataLoader implements CommandLineRunner {

    private static final Logger log = LoggerFactory.getLogger(ScoreDataLoader.class);

    private final ScoreSkillRepository scoreSkillRepository;
    private final Map<String, Double> statWeights = new ConcurrentHashMap<>();

    public ScoreDataLoader(ScoreSkillRepository scoreSkillRepository) {
        this.scoreSkillRepository = scoreSkillRepository;
    }

    @Override
    public void run(String... args) throws Exception {
        ClassPathResource skillsResource = new ClassPathResource("score_skills.csv");
        ClassPathResource effectsResource = new ClassPathResource("score_effects.csv");
        ClassPathResource weightsResource = new ClassPathResource("stat_weights.csv");

        if (!skillsResource.exists() || !effectsResource.exists()) {
            log.warn("score_skills.csv or score_effects.csv not found on classpath; skipping score seed load.");
            return;
        }

        List<ScoreSkill> skills;
        try (Reader skillsReader = utf8Reader(skillsResource);
             Reader effectsReader = utf8Reader(effectsResource)) {
            skills = readScoreSkills(skillsReader, effectsReader);
        }

        if (skills.isEmpty()) {
            log.warn("Score seed CSV files are empty; no score data loaded.");
        } else if (scoreSkillRepository != null) {
            scoreSkillRepository.deleteAll();
            scoreSkillRepository.saveAll(skills);
            int effectCount = skills.stream().mapToInt(skill -> skill.getEffects().size()).sum();
            log.info("Loaded {} score skills and {} score effects from CSV.", skills.size(), effectCount);
        }

        if (!weightsResource.exists()) {
            log.warn("stat_weights.csv not found on classpath; score stat weights remain empty.");
            return;
        }

        try (Reader weightsReader = utf8Reader(weightsResource)) {
            statWeights.clear();
            statWeights.putAll(readStatWeights(weightsReader));
            log.info("Loaded {} score stat weights from CSV.", statWeights.size());
        }
    }

    public Map<String, Double> getStatWeights() {
        return Collections.unmodifiableMap(statWeights);
    }

    List<ScoreSkill> readScoreSkills(Reader skillsReader, Reader effectsReader) throws IOException {
        Map<String, ScoreSkill> skillMap = readSkillMap(skillsReader);
        attachEffects(skillMap, effectsReader);
        return new ArrayList<>(skillMap.values());
    }

    Map<String, Double> readStatWeights(Reader weightsReader) throws IOException {
        Map<String, Double> weights = new LinkedHashMap<>();
        try (CSVReader csvReader = new CSVReaderBuilder(weightsReader).withSkipLines(1).build()) {
            String[] row;
            int lineNumber = 1;
            while ((row = readNext(csvReader, ++lineNumber, "stat_weights.csv")) != null) {
                if (row.length < 2) {
                    log.warn("Skipping line {} in stat_weights.csv: expected 2 columns but found {}", lineNumber, row.length);
                    continue;
                }
                String stat = normalize(row[0]);
                if (stat.isEmpty()) {
                    continue;
                }
                try {
                    weights.put(stat, Double.parseDouble(normalize(row[1])));
                } catch (NumberFormatException ex) {
                    log.warn("Skipping line {} in stat_weights.csv due to invalid weight: {}", lineNumber, row[1]);
                }
            }
        }
        return weights;
    }

    private Map<String, ScoreSkill> readSkillMap(Reader skillsReader) throws IOException {
        Map<String, ScoreSkill> skillMap = new LinkedHashMap<>();
        try (CSVReader csvReader = new CSVReaderBuilder(skillsReader).withSkipLines(1).build()) {
            String[] row;
            int lineNumber = 1;
            while ((row = readNext(csvReader, ++lineNumber, "score_skills.csv")) != null) {
                if (row.length < 5) {
                    log.warn("Skipping line {} in score_skills.csv: expected 5 columns but found {}", lineNumber, row.length);
                    continue;
                }
                String skillKey = normalize(row[0]);
                if (skillKey.isEmpty()) {
                    continue;
                }
                skillMap.put(skillKey, ScoreSkill.builder()
                        .skillKey(skillKey)
                        .cardType(normalize(row[1]).toUpperCase())
                        .position(normalize(row[2]).toUpperCase())
                        .name(normalize(row[3]))
                        .description(normalize(row[4]))
                        .build());
            }
        }
        return skillMap;
    }

    private void attachEffects(Map<String, ScoreSkill> skillMap, Reader effectsReader) throws IOException {
        try (CSVReader csvReader = new CSVReaderBuilder(effectsReader).withSkipLines(1).build()) {
            String[] row;
            int lineNumber = 1;
            while ((row = readNext(csvReader, ++lineNumber, "score_effects.csv")) != null) {
                if (row.length < 4) {
                    log.warn("Skipping line {} in score_effects.csv: expected at least 4 columns but found {}", lineNumber, row.length);
                    continue;
                }
                String skillKey = normalize(row[0]);
                ScoreSkill skill = skillMap.get(skillKey);
                if (skill == null) {
                    log.warn("Skipping line {} in score_effects.csv: unknown skill_id '{}'", lineNumber, skillKey);
                    continue;
                }
                ScoreEffect effect = ScoreEffect.builder()
                        .stat(normalize(row[1]))
                        .condition(normalize(row[2]).isEmpty() ? "ALWAYS" : normalize(row[2]))
                        .values(normalize(row[3]))
                        .baseStat(row.length > 4 && !normalize(row[4]).isEmpty() ? normalize(row[4]) : null)
                        .skill(skill)
                        .build();
                skill.getEffects().add(effect);
            }
        }
    }

    private String[] readNext(CSVReader csvReader, int lineNumber, String source) throws IOException {
        try {
            return csvReader.readNext();
        } catch (CsvValidationException ex) {
            log.warn("Skipping malformed line {} in {}: {}", lineNumber, source, ex.getMessage());
            return null;
        }
    }

    private Reader utf8Reader(ClassPathResource resource) throws IOException {
        return new BufferedReader(new InputStreamReader(resource.getInputStream(), StandardCharsets.UTF_8));
    }

    private String normalize(String value) {
        return value == null ? "" : value.trim();
    }
}

File: backend/src/main/kotlin/com/example/skillsim/config/ScoreDataLoader.kt
package com.example.skillsim.config

import com.example.skillsim.model.ScoreEffect
import com.example.skillsim.model.ScoreSkill
import com.example.skillsim.repository.ScoreSkillRepository
import com.opencsv.CSVReader
import com.opencsv.CSVReaderBuilder
import com.opencsv.exceptions.CsvValidationException
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader
import java.io.Reader
import java.nio.charset.StandardCharsets
import java.util.concurrent.ConcurrentHashMap
import org.slf4j.LoggerFactory
import org.springframework.boot.CommandLineRunner
import org.springframework.core.io.ClassPathResource
import org.springframework.stereotype.Component

private val log = LoggerFactory.getLogger(ScoreDataLoader::class.java)

@Component
class ScoreDataLoader(
    private val scoreSkillRepository: ScoreSkillRepository,
) : CommandLineRunner {

    private val mutableStatWeights = ConcurrentHashMap<String, Double>()

    val statWeights: Map<String, Double> get() = mutableStatWeights.toMap()

    override fun run(vararg args: String) {
        val skillsResource = ClassPathResource("score_skills.csv")
        val effectsResource = ClassPathResource("score_effects.csv")
        val weightsResource = ClassPathResource("stat_weights.csv")

        if (!skillsResource.exists() || !effectsResource.exists()) {
            log.warn("score_skills.csv or score_effects.csv not found on classpath; skipping score seed load.")
            return
        }

        val skills = skillsResource.utf8Reader().use { skillsReader ->
            effectsResource.utf8Reader().use { effectsReader ->
                readScoreSkills(skillsReader, effectsReader)
            }
        }

        if (skills.isEmpty()) {
            log.warn("Score seed CSV files are empty; no score data loaded.")
        } else {
            scoreSkillRepository.deleteAll()
            scoreSkillRepository.saveAll(skills)
            val effectCount = skills.sumOf { it.effects.size }
            log.info("Loaded {} score skills and {} score effects from CSV.", skills.size, effectCount)
        }

        if (!weightsResource.exists()) {
            log.warn("stat_weights.csv not found on classpath; score stat weights remain empty.")
            return
        }

        weightsResource.utf8Reader().use { weightsReader ->
            mutableStatWeights.clear()
            mutableStatWeights.putAll(readStatWeights(weightsReader))
            log.info("Loaded {} score stat weights from CSV.", mutableStatWeights.size)
        }
    }

    internal fun readScoreSkills(skillsReader: Reader, effectsReader: Reader): List<ScoreSkill> {
        val skillMap = readSkillMap(skillsReader)
        attachEffects(skillMap, effectsReader)
        return skillMap.values.toList()
    }

    internal fun readStatWeights(weightsReader: Reader): Map<String, Double> {
        val weights = LinkedHashMap<String, Double>()
        csvReader(weightsReader).use { reader ->
            reader.eachRow("stat_weights.csv") { row, lineNumber ->
                if (row.size < 2) {
                    log.warn(
                        "Skipping line {} in stat_weights.csv: expected 2 columns but found {}",
                        lineNumber,
                        row.size,
                    )
                    return@eachRow
                }
                val stat = row[0].normalize()
                if (stat.isEmpty()) {
                    return@eachRow
                }
                val weight = row[1].normalize().toDoubleOrNull()
                if (weight == null) {
                    log.warn(
                        "Skipping line {} in stat_weights.csv due to invalid weight: {}",
                        lineNumber,
                        row[1],
                    )
                    return@eachRow
                }
                weights[stat] = weight
            }
        }
        return weights
    }

    private fun readSkillMap(skillsReader: Reader): Map<String, ScoreSkill> {
        val skillMap = LinkedHashMap<String, ScoreSkill>()
        csvReader(skillsReader).use { reader ->
            reader.eachRow("score_skills.csv") { row, lineNumber ->
                if (row.size < 5) {
                    log.warn(
                        "Skipping line {} in score_skills.csv: expected 5 columns but found {}",
                        lineNumber,
                        row.size,
                    )
                    return@eachRow
                }
                val skillKey = row[0].normalize()
                if (skillKey.isEmpty()) {
                    return@eachRow
                }
                skillMap[skillKey] = ScoreSkill(
                    skillKey = skillKey,
                    cardType = row[1].normalize().uppercase(),
                    position = row[2].normalize().uppercase(),
                    name = row[3].normalize(),
                    description = row[4].normalize(),
                )
            }
        }
        return skillMap
    }

    private fun attachEffects(skillMap: Map<String, ScoreSkill>, effectsReader: Reader) {
        csvReader(effectsReader).use { reader ->
            reader.eachRow("score_effects.csv") { row, lineNumber ->
                if (row.size < 4) {
                    log.warn(
                        "Skipping line {} in score_effects.csv: expected at least 4 columns but found {}",
                        lineNumber,
                        row.size,
                    )
                    return@eachRow
                }
                val skillKey = row[0].normalize()
                val skill = skillMap[skillKey]
                if (skill == null) {
                    log.warn(
                        "Skipping line {} in score_effects.csv: unknown skill_id '{}'",
                        lineNumber,
                        skillKey,
                    )
                    return@eachRow
                }
                val condition = row[2].normalize().ifEmpty { "ALWAYS" }
                val baseStat = row.getOrNull(4)?.normalize()?.ifEmpty { null }
                skill.effects += ScoreEffect(
                    stat = row[1].normalize(),
                    condition = condition,
                    values = row[3].normalize(),
                    baseStat = baseStat,
                    skill = skill,
                )
            }
        }
    }

    private fun csvReader(reader: Reader): CSVReader =
        CSVReaderBuilder(reader).withSkipLines(1).build()

    /**
     * 헤더를 건너뛴 뒤 남은 행을 하나씩 넘긴다.
     *
     * 깨진 행은 예외로 올리지 않고 경고만 남기고 멈춘다. 데이터 한 줄 때문에 부팅이
     * 죽는 것보다 낫다는 판단이 원본 로직에 담겨 있었다.
     */
    private inline fun CSVReader.eachRow(source: String, action: (List<String>, Int) -> Unit) {
        var lineNumber = 1
        while (true) {
            lineNumber++
            val row = try {
                readNext()
            } catch (ex: CsvValidationException) {
                log.warn("Skipping malformed line {} in {}: {}", lineNumber, source, ex.message)
                return
            } catch (ex: IOException) {
                log.warn("Stopped reading {} at line {}: {}", source, lineNumber, ex.message)
                return
            } ?: return
            action(row.toList(), lineNumber)
        }
    }

    private fun ClassPathResource.utf8Reader(): Reader =
        BufferedReader(InputStreamReader(inputStream, StandardCharsets.UTF_8))

    private fun String?.normalize(): String = this?.trim() ?: ""
}
